standard/expressions.md

```diff
@@ -2992,6 +3054,88 @@ A *default_value_expression* is a constant expression ([§12.23](expressions.md#
+    ;
+
+stackalloc_initializer
+     : '{' stackalloc_initializer_element_list '}'
+     ;
+
+stackalloc_initializer_element_list
+     : stackalloc_element_initializer (',' stackalloc_element_initializer)* ','?
+     ;
+    
+stackalloc_element_initializer
+    : expression
+    ;
+```
+
+The *unmanaged_type* ([§8.8](types.md#88-unmanaged-types)) indicates the type of the items that will be stored in the newly allocated location, and the *expression* indicates the number of these items. Taken together, these specify the required allocation size. As the size of a stack allocation cannot be negative, it is a compile-time error to specify the number of items as a *constant_expression* that evaluates to a negative value.
+
+If *unmanaged_type* is omitted, it is inferred from the corresponding *stackalloc_initializer*. If *expression* is omitted from *stackalloc_expression*, it is inferred to be the number of *stackalloc_element_initializer*s in the corresponding *stackalloc_initializer*.
+
+When a *stackalloc_expression* includes both *expression* and *stackalloc_initializer*, the *expression* shall be a *constant_expression* and the number of elements in that *stackalloc_initializer* shall match the value of *expression*.
+
+A stack allocation initializer of the form `stackalloc T[E]` requires `T` to be an *unmanaged_type* and `E` to be an expression implicitly convertible to type `int`. The operator allocates `E * sizeof(T)` bytes from the call stack. The result is a pointer, of type `T*`, to the newly allocated block. For use in safe contexts, a *stackalloc_expression* has an implicit conversion from `T*` to `Span<T>`. As pointer contexts require unsafe code, see §stack-allocation for more information.
+
+If `E` is a negative value, then the behavior is undefined. If `E` is zero, then no allocation is made, and the value returned is implementation-defined. If there is not enough memory available to allocate a block of the given size, a `System.StackOverflowException` is thrown.
+
+When *stackalloc_initializer* is present, the *stackalloc_initializer_element_list* shall consist of a sequence of expressions, each having an implicit conversion to *unmanaged_type* ([§10.2](conversions.md#102-implicit-conversions)). The expressions initialize elements in the allocated memory in increasing order, starting with the element at index zero. In the absence of a *stackalloc_initializer*, the content of the newly allocated memory is undefined.
+
+Access via an instance of `System.Span<T>` to the elements of an allocated block is range checked.
+
+Stack allocation initializers are not permitted in `catch` or `finally` blocks ([§13.11](statements.md#1311-the-try-statement)).
+
+> *Note*: Stack allocation initializers are allowed in `async` methods, but their return value can't be assigned. Neither pointers nor `ref struct` types, like `Span<T>` are allowed in `async` methods. *end note*
```

Comment:
>but their return value can't be assigned [](http://example.com/codeflow?start=72&length=40)

What does this mean?

Reply:
Admittedly not great wording. It's meant to clarify this situation:

```csharp
// This is legal, stackalloc result not assigned to a local variable.
public async Task<int> M4()
{
    _ = stackalloc int[3];
    await Task.Delay(1000);
    return 5;
}

// This is not legal. stackalloc result assigned to local variable, and captured in state machine:
public async Task<int> M4()
{
    Span<int> storage = stackalloc int[3]; // error can't declare a ref struct in an async method
    await Task.Delay(1000);
    return 5;
}

// This is legal. Stackalloc result passed to synchronous method, but not assigned to local variable,
// not captured in state machine:
async Task<int> M4()
{
    int r = M5(stackalloc int[3]);
    await Task.Delay(1000);
    return 5;
}

int M5(Span<int> arg)
{
    return arg.Length;
}
```

Reply:
Thanks for clarifying. I thought "their return value" referred to the return value of the `async` method. Would it make sense to mention the stack allocation explicitly? Perhaps "but the stack allocated result cannot be assigned".